text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+part of the MIR itself. It does however present an interesting option
+for (eventually) stabilizing access to the compiler's internals.
+ 
+[proto-crate]: https://github.com/nikomatsakis/rust/tree/mir/src/librustc_mir
+[repr]: https://github.com/nikomatsakis/rust/blob/mir/src/librustc_mir/repr.rs
+[build]: https://github.com/nikomatsakis/rust/tree/mir/src/librustc_mir/build
+[hir]: https://github.com/nikomatsakis/rust/blob/mir/src/librustc_mir/hir.rs
+[1191]: https://github.com/rust-lang/rfcs/pull/1191
+[tcx]: https://github.com/nikomatsakis/rust/blob/mir/src/librustc_mir/tcx/mod.rs
+
+### Overview of the MIR
+
+The proposed MIR always describes the execution of a single fn.  At
+the highest level it consists of a series of declarations regarding
+the stack storage that will be required and then a set of basic
+blocks:
+
+    MIR = fn({TYPE}) -> TYPE {
+        {let [mut] B: TYPE;}  // user-declared bindings and their types
+        {let TEMP: TYPE;}     // compiler-introduced temporary
+        {BASIC_BLOCK}         // control-flow graph
+    };
+    
+The storage declarations are broken into two categories. User-declared
+bindings have a 1-to-1 relationship with the variables specified in
+the program. Temporaries are introduced by the compiler in various
+cases. For example, borrowing an lvalue (e.g., `&foo()`) will
+introduce a temporary to store the result of `foo()`. Similarly,
+discarding a value `foo();` is translated to something like `let tmp =
+foo(); drop(tmp);`).  Temporaries are single-assignment, but because
+they can be borrowed they may be mutated after this assignment and
+hence they differ somewhat from variables in a pure SSA
+representation.
+
+The proposed MIR takes the form of a graph where each node is a *basic
+block*. A basic block is a standard compiler term for a continuous
+sequence of instructions with a single entry point.  All interesting
+control-flow happens between basic blocks. Each basic block has an id
+`BB` and consists of a sequence of statements and a terminator:
+
+    BASIC_BLOCK = BB: {STATEMENT} TERMINATOR
+    
+A `STATEMENT` can have one of three forms:
+
+    STATEMENT = LVALUE "=" RVALUE        // assign rvalue into lvalue
+              | Drop(DROP_KIND, LVALUE)  // drop value if needed
+    DROP_KIND = SHALLOW                  // (see discussion below)
+              | DEEP
+             
+The following sections dives into these various kinds of statements in
+more detail.
+
+The `TERMINATOR` for a basic block describes how it connects to
+subsequent blocks:
+
+    TERMINATOR = GOTO(BB)              // normal control-flow
+               | PANIC(BB)             // initiate unwinding, branching to BB for cleanup
+               | IF(LVALUE, BB0, BB1)  // test LVALUE and branch to BB0 if true, else BB1
+               | SWITCH(LVALUE, BB...) // load discriminant from LVALUE (which must be an enum),
+                                       // and branch to BB... depending on which variant it is
```

Comment:
@eefriedman 

> Any particular reason to choose SWITCH over some sort of RVALUE which just extracts the discriminant of an enum?

Not really. Mainly that the range of values that switch operates over is easily computed and if we want to check downcasts that are guarded by a switch it is very straightforward. But it's easily possible in either case, and it'd be consistent with `LEN`.

@Aatch 

> I think the SWITCH here is too limited. From the description, it looks like you need to account for every single variant in the enum, even if the match only had (for example) one matching arm and a default. I also think it should be usable for non-enum values too, at least scalar values.

Yes, I'm amenable to tweaking this part. The important thing I wanted to get at is that we deconstruct matches -- the precise set of blocks we use for that (and how we do it) is less important. Regarding the SWITCH I described, it does require N exits, but then you can direct them all to one place if you choose. And yes, matching against scalar values would be nice, right now the matches deconstruct to a series of ifs (though LLVM, it seems, is likely to optimize those later?)